SDWebImage/SDWebImageDownloader.m

```diff
@@ -429,20 +429,20 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didRece
 @implementation SDWebImageDownloadToken
 
 - (void)dealloc {
-    [[NSNotificationCenter defaultCenter] removeObserver:self name:SDWebImageDownloadReceiveResponseNotification object:nil];
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:SDWebImageDownloadReceiveResponseNotification object:_downloadOperation];
 }
 
-- (instancetype)init {
-    self = [super init];
-    if (self) {
-        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(downloadReceiveResponse:) name:SDWebImageDownloadReceiveResponseNotification object:nil];
+- (void)setDownloadOperation:(NSOperation<SDWebImageDownloaderOperation> *)downloadOperation {
+    if (downloadOperation != _downloadOperation) {
+        [[NSNotificationCenter defaultCenter] removeObserver:self name:SDWebImageDownloadReceiveResponseNotification object:_downloadOperation];
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(downloadReceiveResponse:) name:SDWebImageDownloadReceiveResponseNotification object:downloadOperation];
+        _downloadOperation = downloadOperation;
     }
```

Comment:
I'm not sure if messing with the observers doesn't create any issues. It's not usual to keep adding and removing observers this way.

Reply:
Also, the `SDWebImageDownloadReceiveResponseNotification` notification is a generic notification, which means the communication is 1-to-many. Most of the notification patterns involve needing to filter some of the notifications received: for example, observing the changes for a certain keypath.
I suggest we keep things as they are, clearer and without errors.

I do appreciate your improvement, but I feel we don't really need it.

Reply:
@bpoplauschi 🤔 Maybe I think it's no issues, and we also use this way in code, please see [code](https://github.com/rs/SDWebImage/blob/430414eb4bc26ae2c7d77040a535e589ab5e9b86/SDWebImage/SDWebImageDownloader.m#L63-L71), secondly, `SDWebImageDownloadToken` and `DownloaderOperation` is 1-to-1, the token only interested in corresponding downloaderOperation. In my way, we still has filter, but we put this filter in `NSNotificationCenter`.
But anyway, if you guys think we don't need to do this, I can close it. 😂

Reply:
Let's hear another opinion before deciding. The [SDWebImageDownloader initialize](https://github.com/rs/SDWebImage/blob/430414eb4bc26ae2c7d77040a535e589ab5e9b86/SDWebImage/SDWebImageDownloader.m#L63-L71) example is different, there we just do that once for the entire class

Reply:
If anyone wants to get `SDWebImageDownloadReceiveResponseNotification`, he can just set `setDownloadOperation` to nil, everything works fine.

Reply:
👌 Waiting for other guys's opinion. TBO, `downloadOperation` property is defined in internal, and also do once when created, please see [code](https://github.com/rs/SDWebImage/blob/430414eb4bc26ae2c7d77040a535e589ab5e9b86/SDWebImage/SDWebImageDownloader.m#L293-L296).

Reply:
This seems a good way to filter the operation. But the zero weak reference in Objective-C may be not as you think from their implementation. What if the weak object during `addObserver:selector:name:object:]` is alive, but when later a notification is triggered and the weak object is already dealloced (Like KVO, which will crash for this case) ? This need some detailed behavior check and maybe not so obvious to answer.

Reply:
And. Since you say `also do once when created`. What about to just create a init method like this, to init the download operation once and does not change it ?

```objective-c
@implementation SDWebImageDownloadToken

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:SDWebImageDownloadReceiveResponseNotification object:self.downloadOperation];
}

- (instancetype)initWithDownloadOperation:(NSOperation<SDWebImageDownloaderOperation> *)downloadOperation {
    self = [super init];
    if (self) {
        self.downloadOperation = downloadOperation;
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(downloadReceiveResponse:) name:SDWebImageDownloadReceiveResponseNotification object:downloadOperation];
    }
    return self;
}

- (void)downloadReceiveResponse:(NSNotification *)notification {
    NSOperation<SDWebImageDownloaderOperation> *downloadOperation = notification.object;
    if (downloadOperation && downloadOperation == self.downloadOperation) {
        self.response = downloadOperation.response;
    }
}
@end
```

Reply:
> What if the weak object during addObserver:selector:name:object:] is alive, but when later a notification is triggered and the weak object is already dealloced (Like KVO, which will crash for this case) ?

@dreampiggy No notification would be post, `NSNotificationCenter` would not retain observer and object, why we should remove observer before is `Apple` keep the observer as `unsafe_unretaind`, not `weak` (`weak` is introduced in `ARC`), after `iOS9`, we don't need to remove observer in code anymore.
And for your consideration, let's assume object is weak, after added observer, it becomes nil, that is the object dealloc, other words, sender is gone, no notification would post again.

Reply:
> And. Since you say also do once when created. What about to just create a init method like this, to init the download operation once and does not change it ?

@dreampiggy , I updated the commit, remove the setter of `downloadOperation`, please see the code.

Reply:
@zhongwuzw Yeah...If the `object` arg in the `addObserver:selector:name:object:` was not be stored as `unsafe_ unsafe_unretaind` like the behavior of `observer` before iOS 9, so my worry can be disappear.

I test this code with demo running iOS 8, it seems works without any issue. Maybe this can be adoptable.

Reply:
@dreampiggy Yeah, just like my comments above, you can simply assume `NSNotificationCenter` has a map, the key is the address of `object`, it acts as a route key, only sender's key match this key, dispatch would be happen.

Reply:
And apart from internal implementation, let's just see `Apple`'s doc:

> anObject
The object whose notifications the observer wants to receive; that is, only notifications sent by this sender are delivered to the observer.
If you pass nil, the notification center doesn’t use a notification’s sender to decide whether to deliver it to the observer.